### DISCUSSION TITLE
Writing entire structs

### DISCUSSION THREAD

**Discussion by andrewwong0893** [fisothemes/TwinCAT-ADS-Guide-in-.NET#3] [Q&A]
Hello, I have read your blog on LinkedIn and have found it interesting and the different approaches that are out there.

I am curious why you did not go with the following methods as I believe them to be much simpler for reading and writing variables anywhere in the PLC program.
https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_ads.net/9407898379.html&id=7623751036671403167
![image](https://github.com/user-attachments/assets/941df67d-4ca3-4231-8c03-a26756370fc7)

https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_ads.net/9407799051.html&id=2248463880776176461
![image](https://github.com/user-attachments/assets/50fcfa51-61bf-4814-9b3f-1409aa6e25b2)

This works for all primitive types, arrays, structs, property inside of a struct and arrays of structs.

If you have a variable that you want to access that is deep inside multiple layers of function blocks, your symbol path would simply be:
```
MAIN.fb1.fb2.fb3....fbn.myVariable
```

If you are lazy, can you log into your PLC, navigate to your variable, right click on the variable and copy. This will copy the entire path of that variable/symbol

As for structs, the setup is fairly complicated.
First, you design your model as a struct type in dotNet with the attributes [StructLayout(LayoutKind.Sequential, Pack = 0)]

```csharp
using System.Runtime.InteropServices;

[StructLayout(LayoutKind.Sequential, Pack = 0)]
public struct MyModelPlcExample
{
    public short MyPlcInt;

    public float MyPlcReal;

    public double MyPlcLreal;

    [MarshalAs(UnmanagedType.I1)]
    public bool MyPlcBool;

    [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 21)]
    public string MyPlcStringOfSize20;

    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 10)]
    public byte[] MyPlcByteArrayFromOneToTen;

    [MarshalAs(UnmanagedType.ByValArray, ArraySubType = UnmanagedType.I1, SizeConst = 5)]
    public bool[] MyPlcBoolArrayFromOneToFive;
}
```

The PLC Struct
```
TYPE ST_MyModelPlcExample:
STRUCT
    nMyInt : INT;

    fMyReal : REAL;

    fMyLRreal : LREAL;

    bMyBool : BOOL;

    sMyString : STRING[20];

    aMyByteArray : ARRAY[1..10] OF BYTE;

    aMyBoolArray : ARRAY[1..5] OF BOOL;
END_STRUCT
END_TYPE
```

Noticed that:
- the order of the variables MATTERS
- ADS transfers all messages as a series of bytes so a bool needs to be converted to a byte
- Array sizes needs to be defined
- Strings are always size defined in the PLC + 1

If my struct is just in MAIN, then your C# call is simply:

```csharp

var structToWrite = new MyModelPlcExample ... (set up your values like a constructor);

await adsClient.WriteValueAsync("MAIN.stMyStruct", structToWrite, CancellationToken.None);
```

**Comment by fisothemes** [comment]
Hi @andrewwong0893,  

Thank you for raising this point, and I truly appreciate the level of detail you’ve put into this issue! 👍  

You’re absolutely right that the `AdsClient.WriteValue<T>(...)` and `AdsClient.ReadValue<T>(...)` methods are simpler approaches for reading and writing PLC symbol values, no arguments there. I’ve used a similar strategy in my [LabVIEW implementation of the TwinCAT ADS library](https://github.com/fisothemes/LabVIEW-TwinCAT-ADS). However, as you’ve correctly noted, these methods in Beckhoff's official TwinCAT ADS library for .NET only support **primitive types and 1D arrays of primitive types**. Once you step beyond that, things get tricky. 

Here are some issues I’ve encountered:  


### 1. **Silent Data Corruption**  

When a struct is modified in the PLC (e.g., changing a data type) but the overall byte size remains the same, value marshalling won’t detect this. This can cause silent data corruption, leading to unexpected behaviour.  

#### Example: Changing an `INT` to Two `BOOL` Flags  

**Original PLC Struct:**  
```js
TYPE ST_MyStruct:
STRUCT
    nValue : INT;
END_STRUCT
END_TYPE
```  

**Modified PLC Struct:**  
```js
TYPE ST_MyStruct:
STRUCT
    bFlag1 : BOOL := TRUE;
    bFlag2 : BOOL := TRUE;
END_STRUCT
END_TYPE
```  

**C# Value-Marshalled Struct:**  
```cs
[StructLayout(LayoutKind.Sequential, Pack = 8)]
public struct MyStruct
{
    public short Value; // Still assumes INT (2 bytes)
}
```  

With value marshalling, this code runs fine:  
```cs
var myStruct = client.ReadValue<MyStruct>("MAIN.stMyStruct");
Console.WriteLine(myStruct.Value);
```  

However, with dynamic marshalling, you’ll get a `RuntimeBinderException` like this:  
```cs
dynamic myStruct = MAIN.stMyStruct.ReadValue();
Console.WriteLine(myStruct.nValue); // Throws: 'DynamicValue' does not contain a definition for 'nValue'
```  

Dynamic marshalling would pick up the structural change immediately, safeguarding against silent errors. It would let you access the new fields:  
```cs
Console.WriteLine(myStruct.bFlag1);
```  

### 2. **Hardware Dependency**  

Value marshalling ties your .NET application to the Beckhoff device’s CPU architecture because of struct padding. On x64 devices, you need `Pack = 8`, and on x86, it’s `Pack = 4`. The usual workaround is to set `Pack = 0` or `Pack = 1` and use the PLC pragma `{attribute 'pack_mode' := '0'}` to eliminate padding on both sides.  

But this “solution” has two major problems:  

1. **Incompatibility with ARM CPUs:** CX devices with ARM processors, like the CX9020, don’t support `pack_mode := '0'`.  
2. **Native/Library Structs:** Non-packed structs from PLC libraries (native or custom) become inaccessible.  

### 3. **Complex Nested Structs**  

The deeper the nesting of structs, or the more arrays of structs you include, the more unwieldy value marshalling becomes:  

- C# struct definitions grow exponentially.  
- Any small PLC struct change requires manual updates in C#, increasing maintenance overhead.  
- Aligning each layer of the struct layout adds risk, making the system fragile.  

For example, consider this relatively simple nested struct:  

**PLC Struct:**  
```js
TYPE ST_Node:
STRUCT
    arData  : ARRAY[0..3] OF ARRAY[0..3] OF ST_Data;
    pNext   : POINTER TO ST_Node;
END_STRUCT
END_TYPE
```  
```js
PROGRAM MAIN
VAR
    stNode : ST_Node := (pNext := ADR(stNode));
END_VAR
```  

Value marshalling this structure would be tedious and error-prone. But with dynamic marshalling, it’s straightforward:  

```cs
// Access the current node’s data
dynamic current = MAIN.stNode.ReadValue();
Console.WriteLine(JsonConvert.SerializeObject(current.arData));

// Access the next node’s data
dynamic next = MAIN.stNode.pNext.Reference.ReadValue();
Console.WriteLine(JsonConvert.SerializeObject(next.arData));
```  

### 4. **ReadValue and WriteValue Fail with Arrays**  

The `ReadValue<T>(...)` and `WriteValue<T>(...)` methods fail outright if you try to read or write multi-dimensional arrays. Instead, you have to manage handles manually, which introduces complexity, risks like resource leaks, and invalid resource accesses.  

### On Transferring Structs as Bytes  

You’re absolutely correct that dynamic marshalling struggles to transfer a struct as a series of byte, member mutation requires multiple calls. This is something I intentionally left open in my guide, hoping someone might contribute a solution. It appears Beckhoff has implemented such functionality, but unfortunately, their documentation doesn’t cover it. You might want to contact Beckhoff Technical Support for clarification.  

For now, my workaround is to read the struct, modify it in memory, and write it back. While not ideal for high-performance scenarios, it works reliably. If performance is critical, consider caching or using the factory pattern to minimise repeated accesses.

Here’s an example of my approach:  

```cs
dynamic myStruct = MAIN.stMyStruct.ReadValue();
myStruct.Flag2 = false;
MAIN.stMyStruct.WriteValue(myStruct);
```  

### Why I Use Dynamic Marshalling  

Dynamic marshalling isn’t just about being lazy (though I admit it does make life easier!). It’s a **more robust and flexible solution** that lets you focus on solving real problems instead of managing fragile structures and layouts.  

For example:  

- **No silent data corruption:** Changes to the PLC struct are caught at runtime.  
- **Cross-platform compatibility:** It works across all Beckhoff hardware, regardless of CPU architecture.  
- **Ease of use:** It’s faster to implement and maintain.  

Thanks again for this detailed issue! If you have any other thoughts or suggestions, feel free to share them.
